src/Dictionaries/CacheDictionary.cpp

```diff
@@ -805,42 +881,14 @@ void CacheDictionary::waitForCurrentUpdateFinish(UpdateUnitPtr & update_unit_ptr
 
     if (!result)
     {
-        std::lock_guard<std::mutex> callback_lock(update_unit_ptr->callback_mutex);
-        /*
-         * We acquire a lock here and store false to special variable to avoid SEGFAULT's.
-         * Consider timeout for wait had expired and main query's thread ended with exception
-         * or some other error. But the UpdateUnit with callbacks is left in the queue.
-         * It has these callback that capture god knows what from the current thread
-         * (most of the variables lies on the stack of finished thread) that
-         * intended to do a synchronous update. AsyncUpdate thread can touch deallocated memory and explode.
-         * */
-        update_unit_ptr->can_use_callback = false;
         throw DB::Exception(ErrorCodes::TIMEOUT_EXCEEDED,
                             "Dictionary {} source seems unavailable, because {}ms timeout exceeded.",
                             getDictionaryID().getNameForLogs(), toString(query_wait_timeout_milliseconds));
     }
 
 
     if (update_unit_ptr->current_exception)
-    {
-        // There might have been a single update unit for multiple callers in
-        // independent threads, and current_exception will be the same for them.
-        // Don't just rethrow it, because sharing the same exception object
-        // between multiple threads can lead to weird effects if they decide to
-        // modify it, for example, by adding some error context.
-        try
-        {
-            std::rethrow_exception(update_unit_ptr->current_exception);
-        }
-        catch (...)
-        {
-            throw DB::Exception(ErrorCodes::CACHE_DICTIONARY_UPDATE_FAIL,
-                "Update failed for dictionary '{}': {}",
-                getDictionaryID().getNameForLogs(),
-                getCurrentExceptionMessage(true /*with stack trace*/,
-                    true /*check embedded stack trace*/));
-        }
-    }
+        std::rethrow_exception(update_unit_ptr->current_exception);
```

Comment:
This code is prone to the same problem with exception copying here. I think it would help to make this place more defensive -- either consume the `current_exception` (zero it), or copy it.

Reply:
Thanks.